# I. Import Library and Data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
org_f302 = pd.read_csv(r"D:\Data\3. FTU\Vòng 3\[G'CONTEST 2024] DATABASE\final\f302.csv")
f302 = org_f302.copy()
f302.head()

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_18608\1414480439.py:1: DtypeWarning: Columns (4,13,14,30,31) have mixed types. Specify dtype option on import or set low_memory=False.
  org_f302 = pd.read_csv(r"D:\Data\3. FTU\Vòng 3\[G'CONTEST 2024] DATABASE\final\f302.csv")


BACKUP_DATE  BRANCH_CODE     CIFBRANCH  CUST_CLASSIFICATION  \
0  2020-12-31            1  01809513-001             100200.0   
1  2020-12-31            1  00295876-001             100100.0   
2  2020-12-31            1  00616063-001             100100.0   
3  2020-12-31            1  01573562-001             100100.0   
4  2020-12-31            1  02027435-001             100100.0   

  CUSTOMER_CATEGORY     GLCODE  CCY  ACY_CURR_BALANCE  LCY_CURR_BALANCE  \
0           CA NHAN  213100000  VND      8.661258e+06      8.661258e+06   
1           CA NHAN  213100000  VND      1.515511e+07      1.515511e+07   
2           CA NHAN  213100000  VND      4.077064e+06      4.077064e+06   
3           CA NHAN  213100000  VND      2.290855e+06      2.290855e+06   
4           CA NHAN  213300000  VND      3.138829e+06      3.138829e+06   

   VALUE_DATE  ... ODIP.1  CIC  DEBT SOLD CARD MAXLOAN CIFMAX INT RECAL  \
0  2018-05-24  ...    1.0  0.0        0.0  NaN     1.0      1       0.0   
1  2018-05-25  ...    1.0  0.0        0.0  0.0     1.0      1   26334.0   
2  2018-05-30  ...    1.0  0.0        0.0  NaN     1.0      1       0.0   
3  2018-05-31  ...    1.0  0.0        0.0  1.0     1.0      1    4480.0   
4  2018-05-31  ...    3.0  0.0        0.0  0.0     3.0      3       0.0   

   Confirmation sent CIF_generated  CUST_AC_NO_generated  
0                NaN       53450.0                  1162  
1                NaN       10049.0                  1163  
2                NaN       17078.0                  1164  
3                NaN       38088.0                  1165  
4                NaN       79215.0                  1166  

[5 rows x 43 columns]

In [6]:
org_f303 = pd.read_csv(r"D:\Data\3. FTU\Vòng 3\[G'CONTEST 2024] DATABASE\final\f303.csv")
f303 = org_f303.copy()
f303.head()

REASON_TYPE STOPACC COVID         Số TK vay Ngày giải ngân Ngày đáo hạn  \
0       COVID       Y     x  001LHC1133570001     2013-12-23   2022-12-23   
1       COVID       Y     x  001LHC1151570001     2015-06-06   2025-06-06   
2       COVID       Y     x  001LVND153640001     2015-12-30   2025-12-30   
3       COVID       Y     x  001LVND160600002     2016-02-29   2026-02-28   
4       COVID       Y     x  001MVND160670501     2016-03-07   2026-03-07   

                            Mục đích vay cụ thể           INITIALS  \
0                           COVID1_VAY TIEU DUNG  001LHC1133570001   
1              COVID1_VAY TIEU DUNG THE CHAP BDS  001LHC1151570001   
2            COVID1_THANH TOAN CHO HD SO 0000070  001LVND153640001   
3  COVID1_TT TIEN THEO HD SO 01 TV&ĐVC 2015/HĐKT  001LVND160600002   
4           COVID1_TT TIEN TAM UNG DOT 1 CHO CTY  001MVND160670501   

   CIF_generated  
0           8528  
1           8528  
2         394361  
3         394361  
4         394361

In [7]:
org_f304 = pd.read_csv(r"D:\Data\3. FTU\Vòng 3\[G'CONTEST 2024] DATABASE\final\f304.csv")
f304 = org_f304.copy()
f304.head()

Nhóm nợ thẻ tín dụng  Nhóm chốt max   Du no chot 31.12.2020  \
0                   1.0            1.0            2.569285e+08   
1                   1.0            1.0            7.137838e+05   
2                   2.0            2.0            1.827520e+07   
3                   1.0            1.0            1.262733e+08   
4                   1.0            1.0            6.791053e+06   

   Lãi quá hạn lũy kế hệ thống  Lãi quá hạn lũy kế đã thu  Lãi quá hạn lũy kế  \
0                     0.000000                   0.000000             0.00000   
1                     0.000000                   0.000000             0.00000   
2                  5642.944659                2383.031389          3259.91327   
3                     0.000000                   0.000000             0.00000   
4                          NaN                        NaN             0.00000   

   Dư nợ tính dự phòng  Tỷ lệ trích lập  \
0         2.569285e+08             0.00   
1         7.137838e+05             0.00   
2         1.827194e+07             0.05   
3                  NaN             0.00   
4         6.791053e+06             0.00   

   Dự phòng cụ thể  phải trích đến 31/12/2020  CIC  DEBT SOLD  ACCMAX  \
0                                    0.000000  0.0          0       1   
1                                    0.000000  0.0          0       1   
2                               913597.037715  0.0          0       2   
3                                    0.000000  0.0          0       1   
4                                    0.000000  NaN          0       1   

   CARDMAX  LOAN  CIFCARD  CIF_generated  
0        1     0        1              1  
1        1     0        1              2  
2        2     0        2              3  
3        1     0        1              4  
4        1     0        1              5

In [8]:
org_f305 = pd.read_csv(r"D:\Data\3. FTU\Vòng 3\[G'CONTEST 2024] DATABASE\final\f305.csv")
f305 = org_f305.copy()
f305.head()

MACIC  NHOMCAONHATTAITCTDKHAC  NHOMNOTAITCTDHOITIN  DUNOTAITCTDHOITIN  \
0  7932520776                       2                    1                 31   
1   133615984                       5                    1                 12   
2  7933221323                       3                    2                 67   
3   132886606                       4                    1                 17   
4   133011221                       2                    1                190   

  TENTCTDCONHOMNOCAONHAT   NGUOIXL MASANPHAM    NGAYSL  CIF_generated  
0                  [309]  AUTO_R18       R18  20201231          13359  
1             [334, 360]  AUTO_R18       R18  20201231          13449  
2                  [311]  AUTO_R18       R18  20201231          13487  
3                  [311]  AUTO_R18       R18  20201231          13507  
4                  [616]  AUTO_R18       R18  20201231          13515

In [10]:
org_f402 = pd.read_csv(r"D:\Data\3. FTU\Vòng 3\[G'CONTEST 2024] DATABASE\final\f402.csv")
f402 = org_f402.copy()
f402.head()

CIF  CUST_AC_NO  CCY  ACY_CURR_BALANCE  LCY_CURR_BALANCE  VALUE_DATE  \
0  385027      217553  VND      3.613481e+07      3.613481e+07  2018-06-13   
1   84335       78146  VND      1.520749e+09      1.520749e+09  2018-07-14   
2   85434      155149  VND      3.596996e+07      3.596996e+07  2018-07-25   
3   83591       89578  VND      3.925967e+08      3.925967e+08  2018-07-09   
4   84115       67428  VND      2.816612e+08      2.816612e+08  2018-07-16   

  MATURITY_DATE CUSTOMER_TYPE   RATE  STATUS  CIC  CARD  LOAN  DDMAX  
0    2022-06-13             C  22.00       4    0     0     0      4  
1    2028-07-14             I  12.20       4    0     0     0      4  
2    2022-07-25             C  20.00       2    0     0     0      2  
3    2024-07-09             I  12.70       2    0     0     0      2  
4    2024-07-16             I  12.35       2    0     0     0      2

In [11]:
org_f402 = pd.read_csv(r"D:\Data\3. FTU\Vòng 3\[G'CONTEST 2024] DATABASE\final\f502.csv")
f402 = org_f402.copy()
f402.head()

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_15004\2164550271.py:1: DtypeWarning: Columns (10,15,16,18,19) have mixed types. Specify dtype option on import or set low_memory=False.
  org_f402 = pd.read_csv(r"D:\Data\3. FTU\Vòng 3\[G'CONTEST 2024] DATABASE\final\f502.csv")


MACN     CIF                  Loại tài sản        Mã TSĐB LTien  \
0     0   38821         O to-Tu von vay (TVV)   01583309OTO1   VND   
1     0  392828         O to-Tu von vay (TVV)  68866699OTO01   VND   
2     1     378  STK, HDTG- TPB phat hanh-CBV  00002760TKU13   USD   
3     1  381460          May moc thiet bi-TVV  06623244MMC01   USD   
4     1  383727   Hang hoa khac- Hang PTD-TVV   13235566HHA7   USD   

   Giá trị định giá  Số tiền QD VND  Giá trị đảm bảo Ngày hiệu lực  \
0      4.920000e+08    4.920000e+08     3.900084e+08     11/1/2016   
1      9.020000e+09    9.020000e+09     8.118000e+09     8/30/2016   
2      4.031460e+05    9.308238e+09     3.628314e+05      9/1/2020   
3      4.600000e+04    1.062094e+09     3.450000e+04      9/6/2019   
4      2.320009e+06    5.356669e+10     0.000000e+00    12/29/2020   

  Ngày kiểm tra TSDB  ... CIFSHARE2  COLSHARE2 CIFSHARE3 COLSHARE3 CIFSHARE4  \
0          1/27/2017  ...       NaN        NaN       NaN       NaN       NaN   
1                NaN  ...       NaN        NaN       NaN       NaN       NaN   
2                NaN  ...       NaN        NaN       NaN       NaN       NaN   
3                NaN  ...       NaN        NaN       NaN       NaN       NaN   
4                NaN  ...       NaN        NaN       NaN       NaN       NaN   

  COLSHARE4 CIFSHARE5 COLSHARE5 CIFSHARE6 COLSHARE6  
0       NaN       NaN       NaN       NaN       NaN  
1       NaN       NaN       NaN       NaN       NaN  
2       NaN       NaN       NaN       NaN       NaN  
3       NaN       NaN       NaN       NaN       NaN  
4       NaN       NaN       NaN       NaN       NaN  

[5 rows x 32 columns]

# II. Data Preprocessing


## 2.1 F302

In [3]:
f302.head()

BACKUP_DATE  BRANCH_CODE     CIFBRANCH  CUST_CLASSIFICATION  \
0  2020-12-31            1  01809513-001             100200.0   
1  2020-12-31            1  00295876-001             100100.0   
2  2020-12-31            1  00616063-001             100100.0   
3  2020-12-31            1  01573562-001             100100.0   
4  2020-12-31            1  02027435-001             100100.0   

  CUSTOMER_CATEGORY     GLCODE  CCY  ACY_CURR_BALANCE  LCY_CURR_BALANCE  \
0           CA NHAN  213100000  VND      8.661258e+06      8.661258e+06   
1           CA NHAN  213100000  VND      1.515511e+07      1.515511e+07   
2           CA NHAN  213100000  VND      4.077064e+06      4.077064e+06   
3           CA NHAN  213100000  VND      2.290855e+06      2.290855e+06   
4           CA NHAN  213300000  VND      3.138829e+06      3.138829e+06   

   VALUE_DATE  ... ODIP.1  CIC  DEBT SOLD CARD MAXLOAN CIFMAX INT RECAL  \
0  2018-05-24  ...    1.0  0.0        0.0  NaN     1.0      1       0.0   
1  2018-05-25  ...    1.0  0.0        0.0  0.0     1.0      1   26334.0   
2  2018-05-30  ...    1.0  0.0        0.0  NaN     1.0      1       0.0   
3  2018-05-31  ...    1.0  0.0        0.0  1.0     1.0      1    4480.0   
4  2018-05-31  ...    3.0  0.0        0.0  0.0     3.0      3       0.0   

   Confirmation sent CIF_generated  CUST_AC_NO_generated  
0                NaN       53450.0                  1162  
1                NaN       10049.0                  1163  
2                NaN       17078.0                  1164  
3                NaN       38088.0                  1165  
4                NaN       79215.0                  1166  

[5 rows x 43 columns]

### 2.1.2 Quick Overview

In [4]:
f302.describe()

BRANCH_CODE  CUST_CLASSIFICATION        GLCODE  ACY_CURR_BALANCE  \
count  454616.000000        454609.000000  4.546160e+05      4.102230e+05   
mean      383.263473        106280.961525  2.122189e+08      2.389257e+08   
std       233.728763          6652.774833  1.385884e+06      4.233653e+09   
min         0.000000        100000.000000  2.111010e+08      8.631188e-01   
25%       202.000000        100200.000000  2.121000e+08      1.254756e+07   
50%       555.000000        109900.000000  2.121000e+08      2.783348e+07   
75%       555.000000        109900.000000  2.123000e+08      1.365134e+08   
max       999.000000        990000.000000  3.721000e+08      1.584188e+12   

       LCY_CURR_BALANCE           Term           RATE       MIS_CODE  \
count      4.264030e+05  431074.000000  417746.000000  452449.000000   
mean       2.665048e+08       4.346240      28.755856  101002.230041   
std        4.476723e+09       4.662364      18.573245    3622.761713   
min        8.808804e-01       0.000000       0.000000       0.000000   
25%        1.267305e+07       2.000000      12.000000  100000.000000   
50%        2.803070e+07       3.000000      26.400000  100000.000000   
75%        1.400215e+08       5.000000      45.000000  100000.000000   
max        1.584188e+12      25.000000      62.000000  110080.000000   

       CUST_OWNERSHIP  INDUSTRY_CODE  ...      NHOM_NO.1         ODIP.1  \
count   454616.000000  454616.000000  ...  454609.000000  454609.000000   
mean        10.361534     761.980199  ...       1.167304       1.167771   
std          1.925962     143.221170  ...       0.516467       0.516780   
min          2.000000     101.000000  ...       1.000000       1.000000   
25%         11.000000     806.000000  ...       1.000000       1.000000   
50%         11.000000     806.000000  ...       1.000000       1.000000   
75%         11.000000     806.000000  ...       1.000000       1.000000   
max         12.000000     807.000000  ...       5.000000       5.000000   

                 CIC      DEBT SOLD           CARD        MAXLOAN  \
count  425304.000000  454609.000000  434857.000000  454609.000000   
mean        0.055375       0.000106       0.134718       1.197541   
std         0.403944       0.014531       0.408960       0.574685   
min         0.000000       0.000000       0.000000       1.000000   
25%         0.000000       0.000000       0.000000       1.000000   
50%         0.000000       0.000000       0.000000       1.000000   
75%         0.000000       0.000000       0.000000       1.000000   
max         5.000000       2.000000       5.000000       5.000000   

              CIFMAX     INT RECAL  CIF_generated  CUST_AC_NO_generated  
count  454616.000000  4.546090e+05  454609.000000         454616.000000  
mean        1.200334  1.226852e+06  204871.604273         230100.248029  
std         0.580067  6.589628e+07  114227.040333         132261.774681  
min         1.000000  0.000000e+00       6.000000              1.000000  
25%         1.000000  2.502100e+04  108244.000000         115775.750000  
50%         1.000000  2.586010e+05  204095.000000         230895.500000  
75%         1.000000  7.042340e+05  302291.000000         344624.250000  
max         5.000000  3.733041e+10  397132.000000         458275.000000  

[8 rows x 22 columns]

In [5]:
f302.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 454616 entries, 0 to 454615
Data columns (total 43 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   BACKUP_DATE           454616 non-null  object 
 1   BRANCH_CODE           454616 non-null  int64  
 2   CIFBRANCH             454609 non-null  object 
 3   CUST_CLASSIFICATION   454609 non-null  float64
 4   CUSTOMER_CATEGORY     232817 non-null  object 
 5   GLCODE                454616 non-null  int64  
 6   CCY                   454616 non-null  object 
 7   ACY_CURR_BALANCE      410223 non-null  float64
 8   LCY_CURR_BALANCE      426403 non-null  float64
 9   VALUE_DATE            454609 non-null  object 
 10  MATURITY_DATE         439657 non-null  object 
 11  Term                  431074 non-null  float64
 12  RATE                  417746 non-null  float64
 13  TOD_END_DATE          452449 non-null  object 
 14  MIN_INT_REVN_DATE     452448 non-null  object 
 15  

In [6]:
f302_row, f302_column = f302.shape
print(f"This Table contains: {f302_row} rows")
print(f"This Table contains: {f302_column} column")

This Table contains: 454616 rows
This Table contains: 43 column


In [7]:
f302_null_check = f302.isnull().sum()
f302_null_check

BACKUP_DATE                  0
BRANCH_CODE                  0
CIFBRANCH                    7
CUST_CLASSIFICATION          7
CUSTOMER_CATEGORY       221799
GLCODE                       0
CCY                          0
ACY_CURR_BALANCE         44393
LCY_CURR_BALANCE         28213
VALUE_DATE                   7
MATURITY_DATE            14959
Term                     23542
RATE                     36870
TOD_END_DATE              2167
MIN_INT_REVN_DATE         2168
CUSTOMER_TYPE                7
PRODUCT                   2167
MIS_CODE                  2167
CODE_DESC                 2167
CUST_OWNERSHIP               0
DESC                         0
INDUSTRY_CODE                0
INDUSTRY_DESC                0
Dự thu tồn                   7
NHOM_NO                      8
ACCT                         0
TERM                        18
LOANTYPE                     7
CIFTYPE                      7
INITIALS                     7
COVID-OD                454512
ODIP                    452235
NHOM_NO.

In [8]:
f302_duplicated_check = f302.duplicated().sum()
f302_duplicated_check

np.int64(0)

In [9]:
numerical_cols = f302.select_dtypes(include=['int64', 'float64']).columns
num_numerical = len(numerical_cols)

categorical_cols = f302.select_dtypes(include=['object', 'category']).columns
num_categorical = len(categorical_cols)

print(f"Số cột numerical: {num_numerical}")
print(f"Số cột categorical: {num_categorical}")


Số cột numerical: 22
Số cột categorical: 21


## 2.2 F303

In [ ]:
f303.head()

REASON_TYPE STOPACC COVID         Số TK vay Ngày giải ngân Ngày đáo hạn  \
0       COVID       Y     x  001LHC1133570001     2013-12-23   2022-12-23   
1       COVID       Y     x  001LHC1151570001     2015-06-06   2025-06-06   
2       COVID       Y     x  001LVND153640001     2015-12-30   2025-12-30   
3       COVID       Y     x  001LVND160600002     2016-02-29   2026-02-28   
4       COVID       Y     x  001MVND160670501     2016-03-07   2026-03-07   

                            Mục đích vay cụ thể           INITIALS  \
0                           COVID1_VAY TIEU DUNG  001LHC1133570001   
1              COVID1_VAY TIEU DUNG THE CHAP BDS  001LHC1151570001   
2            COVID1_THANH TOAN CHO HD SO 0000070  001LVND153640001   
3  COVID1_TT TIEN THEO HD SO 01 TV&ĐVC 2015/HĐKT  001LVND160600002   
4           COVID1_TT TIEN TAM UNG DOT 1 CHO CTY  001MVND160670501   

   CIF_generated  
0           8528  
1           8528  
2         394361  
3         394361  
4         394361

### 2.2.1 Quick Overview

In [ ]:
f303_des = f303.describe()

In [ ]:
f303_null_check = f303.isnull().sum()
f303_null_check

REASON_TYPE             1036
STOPACC                 1036
COVID                      0
Số TK vay                  0
Ngày giải ngân             0
Ngày đáo hạn               1
Mục đích vay cụ thể        0
INITIALS                   0
CIF_generated              0
dtype: int64

In [ ]:
f303_duplicated_check = f303.duplicated().sum()
f303_duplicated_check

0

In [ ]:
f303_info = f303.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19287 entries, 0 to 19286
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   REASON_TYPE           18251 non-null  object
 1   STOPACC               18251 non-null  object
 2   COVID                 19287 non-null  object
 3   Số TK vay             19287 non-null  object
 4   Ngày giải ngân        19287 non-null  object
 5   Ngày đáo hạn          19286 non-null  object
 6   Mục đích vay cụ thể   19287 non-null  object
 7   INITIALS              19287 non-null  object
 8   CIF_generated         19287 non-null  int64 
dtypes: int64(1), object(8)
memory usage: 1.3+ MB


- It can be seen that this table contains **19,287 rows** of data, with some columns **having null values**, and **no duplicate** entries.
- Some column names are in Vietnamese and **need to be renamed** accordingly.

In [ ]:
f303_row, f303_column = f303.shape
print(f"This Table contains: {f303_row} rows")
print(f"This Table contains: {f303_column} column")

This Table contains: 19287 rows
This Table contains: 9 column


### 2.2.2 Rename Column


In [ ]:
f303.rename(columns = {'Số TK vay': 'LoanAccountNumber','Ngày giải ngân': 'DisbursementDate','Ngày đáo hạn': 'MaturityDate','Mục đích vay cụ thể ': 'LoanPurposeDetail','INITIALS': 'AccountInitials' }, inplace = True)

### 2.2.3 Data Type Conversion

In [ ]:
f303['DisbursementDate'] = pd.to_datetime(f303['DisbursementDate'], errors = 'coerce')
f303['MaturityDate'] = pd.to_datetime(f303['MaturityDate'], errors = 'coerce')

In [ ]:
# Drop null
f303 = f303.dropna(subset = ['MaturityDate'])

### 2.2.4 Checking Advanced Duplicated


In [ ]:
# Check for duplicates in specific columns
cols_to_check = ['LoanAccountNumber', 'CIF_generated', 'AccountInitials']

for col in cols_to_check:
    answer = f303[col].duplicated().any()
    print(f"Does the column {col} have duplicate values? - {answer}")

Cột LoanAccountNumber có giá trị trùng lặp hay không? - False
Cột CIF_generated có giá trị trùng lặp hay không? - True
Cột AccountInitials có giá trị trùng lặp hay không? - True


- Primary key identified as 'LoanAccountNumber'

In [ ]:
# Check if the values in 'LoanAccountNumber' and 'AccountInitials' are the same
check = f303['LoanAccountNumber'] == f303['AccountInitials']
print(check.value_counts())

False    10435
True      8851
Name: count, dtype: int64

- Conclusion: The suspicion that these two columns are duplicates is not confirmed.


### 2.2.5 Preprocessing 'REASON_TYPE' and 'STOPACC'

In [ ]:
f303 = f303.fillna('NULL')